router/broker.go

```diff
@@ -898,3 +1008,119 @@ func (b *broker) subCountSubscribers(msg *wamp.Invocation) wamp.Message {
 		Arguments: wamp.List{count},
 	}
 }
+
+// syncGetSubHistoryEvents returns all stored events for subscription
+// TODO: Need to filter by authid and/or authrole of caller if stored events have any of them
+// But that's not possible in current arch as meta peer doesn't know anything about caller, just invocation message
+func (b *broker) syncGetSubHistoryEvents(subId wamp.ID) []*storedEvent {
+	if subscription, ok := b.subscriptions[subId]; ok {
+		if storeItem, ok := b.eventHistoryStore[subscription]; ok {
+			var events []*storedEvent
+			for _, event := range storeItem.events {
+				events = append(events, event.event)
+			}
+			return events
+		}
+	}
+
+	return []*storedEvent{}
+}
+
+// eventHistoryLast retrieves N last events for subscription.
+func (b *broker) eventHistoryLast(msg *wamp.Invocation) wamp.Message {
+	var events wamp.List // []*storedEvent
+	subId, ok1 := wamp.AsID(msg.Arguments[0])
+	limit, ok2 := msg.Arguments[1].(int)
+
+	if !ok1 || !ok2 || limit <= 0 {
+		return &wamp.Error{
+			Type:    msg.MessageType(),
+			Request: msg.Request,
+			Details: wamp.Dict{},
+			Error:   wamp.ErrInvalidArgument,
+		}
+	}
+
+	sync := make(chan struct{})
+	b.actionChan <- func() {
+		storedEvents := b.syncGetSubHistoryEvents(subId)
+		filteredEvents := storedEvents[-limit:]
+		events = append(events, filteredEvents)
+		close(sync)
+	}
+	<-sync
+
+	return &wamp.Yield{
+		Request:   msg.Request,
+		Arguments: events,
+	}
+}
+
+// eventHistorySince retrieves events for subscription after specified date.
+func (b *broker) eventHistorySince(msg *wamp.Invocation) wamp.Message {
```

Comment:
I had to do do something similar in a non-open source project recently. I considered using timestamps but instead switched to a simple counter, in order to avoid any time skew problems. Would it make sense to allow the clients to ask for events since the last counter they store locally?

Reply:
On my side I used the zero value as the value smaller than all sequence numbers that show up in the protocol.